go/pkg/cosmos/txhistory.go

```diff
@@ -126,92 +162,416 @@ func Messages(msgs []sdk.Msg) []Message {
+	if err := json.Unmarshal([]byte(str), c); err != nil {
+		return errors.Wrapf(err, "error marshaling cursor from %#v", c)
+	}
+	return nil
+}
 
-	// filter out duplicate transactions
-	seen := make(map[string]bool)
-	uniqueTxs := []client.TxSearchResponseResultTxs{}
-	for _, t := range resTxs {
-		if _, ok := seen[*t.Hash]; !ok {
-			uniqueTxs = append(uniqueTxs, t)
-			seen[*t.Hash] = true
+// read tendermint tx history
+func (s *txHistoryState) readTxHistory() (*TxHistory, error) {
+	// if we have cursor, restore state
+	if s.cursorSerialized != "" {
+		if err := s.cursor.decode(s.cursorSerialized); err != nil {
+			return nil, errors.Wrapf(err, "error decoding cursor %s", s.cursorSerialized)
+		} else {
+			// decrement by 1 so we start on last page we requested from tendermint
+			if s.cursor.TendermintSentPage > 0 {
+				s.cursor.TendermintSentPage--
+			}
+			if s.cursor.TendermintRecvdPage > 0 {
+				s.cursor.TendermintRecvdPage--
+			}
 		}
 	}
 
-	// sort descending by block height
-	sort.Slice(uniqueTxs, func(i, j int) bool {
-		if *uniqueTxs[i].Height == *uniqueTxs[j].Height {
-			return *uniqueTxs[i].Index > *uniqueTxs[j].Index
+	// default pageSize to 10
+	if s.pageSize == 0 {
+		s.pageSize = 10
+	}
+
+	sentResult, err := s.nextSenderTxHistoryPage()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading sent txs for %s", s.address)
+	}
+	total := sentResult.Result.TotalCount
+
+	totalSends, err := strconv.ParseUint(total, 10, 64)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing send total %s", total)
+	}
+	s.cursor.TotalSends = totalSends
+
+	recipResult, err := s.nextRecipTxHistoryPage()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading transfer in txs for %s", s.address)
+	}
+
+	total = recipResult.Result.TotalCount
+	totalRecips, err := strconv.ParseUint(total, 10, 64)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing send total %s", total)
+	}
+
+	s.cursor.TotalReceives = totalRecips
+	logger.Infof("got %d sent and %d recvd txs for %s", len(sentResult.Result.Txs), len(recipResult.Result.Txs), s.address)
+
+	txs := make([]Tx, 0, s.pageSize)
+
+	var senderDone, recipDone bool
+	s.sends, s.recips = uniqueify(sentResult.Result.Txs, recipResult.Result.Txs)
+
+	for uint(len(txs)) < s.pageSize && (!senderDone || !recipDone) {
+		logger.Infof("unique we have %d sent and %d recvd", len(s.sends), len(s.recips))
+
+		// need to refill recvd bucket
+		if len(s.recips) == 0 && !recipDone {
+			_, recipDone, err = s.refill(histTxTypeRecipient)
+			if err != nil {
+				return nil, errors.Wrap(err, "error re-filling recips")
+			}
+		}
+
+		// need to refill sent bucket
+		if len(s.sends) == 0 && !senderDone {
+			_, senderDone, err = s.refill(histTxTypeSender)
+			if err != nil {
+				return nil, errors.Wrap(err, "error re-filling sends")
+			}
+		}
+
+		// get the first tx from each type
+		var topOfSent, topOfRecvd uint64
+		if len(s.sends) > 0 && s.sends[0].Height != nil {
```

Comment:
Were you seeing cases of height null? I didn't think tx_search returned any mempool transactions. And if we do see this case, do we want to just throw it away if not valid, or handle it and show as most recent if mempool? Right now it looks like it could just freeze the set of txs